smslibrary/src/main/java/com/eis/smslibrary/SMSMessageParser.java

```diff
@@ -0,0 +1,124 @@
+package com.eis.smslibrary;
+
+import android.content.Context;
+
+import androidx.annotation.NonNull;
+
+import com.eis.communication.MessageHandler;
+
+import it.lucacrema.preferences.PreferencesManager;
+
+/**
+ * Singleton class used to parse String to SMSMessage and back
+ * Uses a strategy to parse messages, so that any user can update it to its preferred parser
+ * By defaults it uses a default strategy, defined by the library
+ *
+ * @author Luca Crema, Alberto Ursino, Marco Mariotto
+ */
+public class SMSMessageParser implements MessageHandler<String, String, SMSMessage> {
+
+    private static final String PREFERENCE_DEFAULT_STRATEGY = "";
+
+    private SMSParseStrategy parseStrategy;
+    private static SMSMessageParser instance;
+
+    /**
+     * Private constructor
+     * @param context The calling context
+     */
+    private SMSMessageParser(Context context) {
+        SMSParseStrategy savedStrategy = getSavedMessageParseStrategy(context);
+        if(savedStrategy == null) parseStrategy = new DefaultSMSParseStrategy();
+        else parseStrategy = savedStrategy;
+    }
+
+    /**
+     * Method to get the only instance of this class
+     * @param context The calling context
+     * @return {@link SMSMessageParser#instance} the only active instance of this class
+     */
+    public static SMSMessageParser getInstance(Context context){
+        if(instance == null)
+            return new SMSMessageParser(context);
+        return instance;
+    }
+
+    /**
+     * Method to update the parse strategy to a custom one. Needs to be called runtime from a valid Context.
+     *
+     * @param context the calling Context
+     * @param parseStrategy custom message parsing
+     */
+    public void setMessageParseStrategy(Context context, @NonNull final SMSParseStrategy parseStrategy) {
```

Comment:
No need for context here, you already asked it in the constructor/getInstance, it's not good looking to have `SMSMessageeParser.getInstance(ctx).setMessageParseStrategy(ctx /* again */, customStrategy)`. Save it in the constructor.